pydantic_ai_slim/pydantic_ai/ui/adapter.py

```diff
@@ -0,0 +1,364 @@
+from ..agent import AbstractAgent, AgentDepsT, AgentRunResult
+from ..builtin_tools import AbstractBuiltinTool
+from ..exceptions import UserError
+from ..messages import ModelMessage
+from ..models import KnownModelName, Model
+from ..output import OutputSpec
+from ..settings import ModelSettings
+from ..toolsets import AbstractToolset
+from ..usage import RunUsage, UsageLimits
+from .event_stream import BaseEventStream, SourceEvent
+
+try:
+    from starlette.requests import Request
+    from starlette.responses import Response
+except ImportError as e:  # pragma: no cover
+    raise ImportError(
+        'Please install the `starlette` package to use `Agent.to_ag_ui()` method, '
+        'you can use the `ag-ui` optional group — `pip install "pydantic-ai-slim[ag-ui]"`'
+    ) from e
+
+
+__all__ = [
+    'BaseAdapter',
+]
+
+
+RunRequestT = TypeVar('RunRequestT')
+"""Type variable for protocol-specific request types."""
+
+MessageT = TypeVar('MessageT')
+"""Type variable for protocol-specific message types."""
+
+EventT = TypeVar('EventT')
+"""Type variable for protocol-specific event types."""
+
+OnCompleteFunc: TypeAlias = Callable[[AgentRunResult[Any]], None] | Callable[[AgentRunResult[Any]], Awaitable[None]]
+"""Callback function type that receives the `AgentRunResult` of the completed run. Can be sync or async."""
+
+
+# State management types
+
+StateT = TypeVar('StateT', bound=BaseModel)
+"""Type variable for the state type, which must be a subclass of `BaseModel`."""
+
+
+@runtime_checkable
+class StateHandler(Protocol):
+    """Protocol for state handlers in agent runs. Requires the class to be a dataclass with a `state` field."""
+
+    # Has to be a dataclass so we can use `replace` to update the state.
+    # From https://github.com/python/typeshed/blob/9ab7fde0a0cd24ed7a72837fcb21093b811b80d8/stdlib/_typeshed/__init__.pyi#L352
+    __dataclass_fields__: ClassVar[dict[str, Field[Any]]]
+
+    @property
+    def state(self) -> Any:
+        """Get the current state of the agent run."""
+        ...
+
+    @state.setter
+    def state(self, state: Any) -> None:
+        """Set the state of the agent run.
+
+        This method is called to update the state of the agent run with the
+        provided state.
+
+        Args:
+            state: The run state.
+
+        Raises:
+            InvalidStateError: If `state` does not match the expected model.
+        """
+        ...
+
+
+@dataclass
+class StateDeps(Generic[StateT]):
+    """Provides AG-UI state management.
+
+    This class is used to manage the state of an agent run. It allows setting
+    the state of the agent run with a specific type of state model, which must
+    be a subclass of `BaseModel`.
+
+    The state is set using the `state` setter by the `Adapter` when the run starts.
+
+    Implements the `StateHandler` protocol.
+    """
+
+    state: StateT
+
+
+@dataclass
+class BaseAdapter(ABC, Generic[RunRequestT, MessageT, EventT, AgentDepsT]):
```

Comment:
Should we call this BaseAdapter, AbstractAdapter, or something else? Should we make it `<X>UIAdapter`?